### ISSUE TITLE
RC circuit with low values of R has poor performance

### ISSUE THREAD

**Issue by samreid** [phetsims/circuit-construction-kit-common#720]
From #678, @kathy-phet  found that turning down the resistance really disturbs performance:

![image](https://user-images.githubusercontent.com/679486/131032538-95b1e8e0-00b3-46e7-bfed-717b912b994b.png)

**Comment by samreid** [comment]
@KatieWoe can you please also have the QA team test this issue in https://phet-dev.colorado.edu/html/circuit-construction-kit-ac/1.0.0-dev.16/phet/circuit-construction-kit-ac_en_phet.html ?

**Comment by Nancy-Salpepi** [comment]
Using MacBook Air 11.4 + chrome:

I noticed an inconsistency when the resistance was changed in pause vs play mode.

I built the circuit shown above and the voltage curve remained smooth and consistent when the resistance was decreased while the sim was in play mode. The power source did not go on fire. When I paused the sim and changed the resistance from 10 to zero, it showed the power source on fire. When I pressed the play button, the fire went away and the voltage curve was smooth and consistent. 

https://drive.google.com/file/d/1bo26iA8OlrUsNsh9XVH1PLFVUfrm7HiH/view?usp=sharing

I also noticed in play mode that the electrons stop moving in the resistor when resistance is set at zero. 

https://drive.google.com/file/d/1ohfaUKYvq_ZB21RmhJIzEpD0le39QZFZ/view?usp=sharing

**Comment by samreid** [comment]
Thanks, I'll move the problems in the preceding comment to 2 new issues.

**Comment by stemilymill** [comment]
Win10 Chrome

In this circuit with low resistance, the voltage chart looked ok but the current chart was glitchy
![low resistance current chart glitchy 2](https://user-images.githubusercontent.com/85511101/131382465-ea60e91b-1b36-4e93-ba63-f5b816a8c9bf.gif)

**Comment by stemilymill** [comment]
Win10 Chrome, second screen

I don't know if this is just the capacitor doing its thing, but here is an instance with a mostly smooth voltage chart and not smooth current chart

![low resistance current chart glitchy 2](https://user-images.githubusercontent.com/85511101/131389419-21e7a46c-6531-4afa-9107-ca4160511fc9.gif)

**Comment by stemilymill** [comment]
Win10 Chrome, RLC screen.

This recording was at a fairly high frame rate and the jerkiness of the animation is **not** due to conversion to gif - I thought I should specify that in this case. However, the apparent smearing of the resistor *is* due to recording software so please ignore that, and in the previous comment as well.

![low resistance low performance 3 35 frames per second](https://user-images.githubusercontent.com/85511101/131398687-a36d9b19-eb1b-4818-93c2-a84109e15141.gif)

**Comment by samreid** [comment]
Thanks for these reports.  I don't see the symptom on my machine--may be performance sensitive.  I published a version here https://phet-dev.colorado.edu/html/circuit-construction-kit-ac/1.0.0-dev.17/phet/circuit-construction-kit-ac_en_phet.html for a potentially related problem in https://github.com/phetsims/circuit-construction-kit-common/issues/727#issuecomment-908693576.  Can you please test that version and see if the problem persists?  If so, can you please try successively higher values of ?capacitorResistance to see if that helps?  The default value is 1E-6.

**Comment by stemilymill** [comment]
The performance was better on dev 17, but I tried the same circuit in dev 20 and it was much worse, as seen below.
Same behavior on win10 chrome and mac chrome. Next I will test with the different values of ?capacitorResistanc if that is still what you'd like me to do.

![giphy dev20](https://user-images.githubusercontent.com/85511101/131894231-f9ba8678-d3f9-4960-9b57-2950bbb1ddff.gif)

**Comment by stemilymill** [comment]
I made this same circuit in dev 17 - 20. In each, the lightbulb and resistor are 0.1 ohms and the capacitor is default.
I will test with different values for ?capacitorResistance and/or ?precision to see if it makes a difference

dev.17

![dev 17 capacitor default](https://user-images.githubusercontent.com/85511101/132060702-e944b920-34b7-4b35-bd77-53a516271e70.gif)


dev.18

![dev 18 capacitor default](https://user-images.githubusercontent.com/85511101/132060731-db95ab80-6270-402c-94cd-0ddc13981fb1.gif)


dev.19

![dev 19 capacitor default](https://user-images.githubusercontent.com/85511101/132060752-79115dae-8e5c-43e2-afb3-bc50de1b1194.gif)


dev.20

![dev 20 capacitor default](https://user-images.githubusercontent.com/85511101/132060768-31030494-b9ba-4873-82f3-ecf44a168207.gif)

**Comment by samreid** [comment]
At today's design meeting, we were wondering if @stemilymill can test this in https://phet-dev.colorado.edu/html/circuit-construction-kit-ac/1.0.0-dev.24/phet/circuit-construction-kit-ac_en_phet.html ?

**Comment by stemilymill** [comment]
Win10 Chrome
I reproduced the same circuit as in above comments in dev.24. Unfortunately, the performance seems even worse than dev.20.


https://user-images.githubusercontent.com/85511101/132870319-8f7fa3dc-9e18-447f-baed-305c7f8061f1.mp4

**Comment by samreid** [comment]
I created a ac/capacitor/light builb/resistor/switch circuit with 0.1 ohms for the light bulb and resistor, and I do not see performance problems on my machine.  @stemilymill can you help narrow down if the problem is platform-specific?  I've been testing dev.24 to try to replicate your results, but new testing should be done in dev.25.

**Comment by stemilymill** [comment]
@samreid using the same circuit as before with dev.25, on dell laptop and Win10 Chrome, the performance becomes much better at around 1.2 - 1.3 ohms, and normal performance at 1.9 - 2 ohms.

Interestingly, apart from an initial spike when the switch closes, the performance is also normal when the lightbulb and resister are both set to zero Ohms.

 I will continue testing on different platforms tomorrow.

Resistance 1.3 Ohms

https://user-images.githubusercontent.com/85511101/133512243-62e4e52d-85c2-4806-97fa-05927b6df68c.mp4


Resistance 0 Ohms

https://user-images.githubusercontent.com/85511101/133513652-021cd12a-c56b-418c-9204-7c63fe414e70.mp4

**Comment by KatieWoe** [comment]
I ran a fuzz test to test performance and I noticed significant slow down after a bit. Though I'm not sure what caused it. It was to the point where I had trouble closing the window.
This was on dev.26

**Comment by stemilymill** [comment]
@samreid I tested multiple platforms on dev.26 with the same circuit as before. With each I first tested with the lightbulb and resistor both set to 0.1 ohms, then both set to 1.0 ohms.

Win10 Chrome:

https://user-images.githubusercontent.com/85511101/134402082-5c5d0f01-3f0a-443f-8516-387071eb49b9.mp4

https://user-images.githubusercontent.com/85511101/134402117-2fe84ab9-12fc-4a1a-87ef-1241b7713faf.mp4


Win10 Firefox:


https://user-images.githubusercontent.com/85511101/134402153-7d585374-1fe4-4de8-8834-09e6dfacc889.mp4


https://user-images.githubusercontent.com/85511101/134402164-16cc73e2-c523-4b96-bfd8-96de2355edf4.mp4


Win10 Edge:



https://user-images.githubusercontent.com/85511101/134402201-8a2c6e93-81ef-4012-b559-f8dacf7d13ec.mp4

https://user-images.githubusercontent.com/85511101/134402219-7b72e9d4-240c-4001-a6d0-4c4738b73eee.mp4


Macbook Chrome:


https://user-images.githubusercontent.com/85511101/134402250-b6e782be-920a-4b43-9428-599d6a351dac.mp4



https://user-images.githubusercontent.com/85511101/134402275-b62c5c7f-b52d-4c11-b797-94c3f682b540.mp4


Macbook Safari:


https://user-images.githubusercontent.com/85511101/134402310-ad20b404-8859-48df-9d70-357e321c12eb.mp4


https://user-images.githubusercontent.com/85511101/134402336-bad208e9-1b89-4fe4-9fb8-0e59bbb3e5f7.mp4


Chromebook:

https://user-images.githubusercontent.com/85511101/134402423-0a31e012-4765-44b4-a093-4e2fc3c8fb4d.mp4



https://user-images.githubusercontent.com/85511101/134402437-990caff3-654c-4a73-bec4-53fb649bc832.mp4

**Comment by samreid** [comment]
I experimented with changing the error threshold to 1E-3 (current value is 1E-5) and this boosted performance significantly on my Mac/Firefox.  However, this leads to numerical errors in the curve:

![image](https://user-images.githubusercontent.com/679486/135124396-1e195687-c613-438c-9280-2e31c12c372f.png)

Here is a patch I used to investiagte:

```diff
Index: js/model/TimestepSubdivisions.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/model/TimestepSubdivisions.js b/js/model/TimestepSubdivisions.js
--- a/js/model/TimestepSubdivisions.js	(revision 5a40dba98cdfaa067dd50908cdaf13ea0ad193f3)
+++ b/js/model/TimestepSubdivisions.js	(date 1632845082178)
@@ -13,10 +13,10 @@
 import ResultSet from './ResultSet.js';
 
 // smallest possible time
-const MIN_DT = 1E-6;
+const MIN_DT_DEFAULT = 1E-6;
 
 //threshold for determining whether 2 states are similar enough; any error less than errorThreshold will be tolerated.
-const ERROR_THRESHOLD = 1E-5;
+const ERROR_THRESHOLD_DEFAULT = 1E-3;
 
 /**
  * @typedef Steppable
@@ -69,6 +69,9 @@
    */
   search( state, steppable, dt, halfStepState ) {
 
+    const MIN_DT = window.minDT || MIN_DT_DEFAULT;
+    const ERROR_THRESHOLD = window.errorThreshold || ERROR_THRESHOLD_DEFAULT;
+
     // if dt is already too low, no need to do error checking
     if ( dt <= MIN_DT ) {
       return { dt: MIN_DT, state: steppable.update( state, MIN_DT ) };

```

@arouinfar can we schedule a mini-meeting to review this kind of problem?

**Comment by samreid** [comment]
@arouinfar and I discussed the problem.  It seems to be related to the accuracy-performance tradeoff.  Our parameters for tuning:

* Error threshold for the timestep subdivisions
* Number of times to run the arbitrary precision solver before falling back to the floating-point solver
* Minimum dt threshold for the timestep subdivisions

Questions for design meeting:
* What is the slowest performance we would find acceptable on our slowest platform?
* How much can we tolerate numerical imprecision (like the high frequency noise) in the curves?
* Any other ideas for how to improve the performance here?

**Comment by samreid** [comment]
From today's design meeting:

> What is the slowest performance we would find acceptable on our slowest platform?

It's difficult to say, we would need to see it live in the simulation.

**Comment by samreid** [comment]
After the fixes in #744, testing on https://phet-dev.colorado.edu/html/circuit-construction-kit-ac/1.0.0-dev.32/phet/circuit-construction-kit-ac_en_phet.html.  Behavior is good with AC-Capacitor-Resistor, even down to 0.1 Ohms.  In one run, turning down to 0.0 ohms crashed the sim, but in another run it still had good behavior & performance.

**Comment by samreid** [comment]
At today's design meeting, we decided to close this and ask QA to keep their eyes peeled during dev testing.
